src/librustdoc/html/static/js/search.js

```diff
@@ -154,13 +389,101 @@ window.initSearch = function(rawSearchIndex) {
             return NO_TYPE_FILTER;
         }
 
-        var valLower = query.query.toLowerCase(),
-            val = valLower,
-            typeFilter = itemTypeFromName(query.type),
-            results = {}, results_in_args = {}, results_returned = {},
-            split = valLower.split("::");
+        userQuery = userQuery.trim();
+        var query = {
+            original: userQuery,
+            userQuery: userQuery.toLowerCase(),
+            length: userQuery.length,
+            pos: 0,
```

Comment:
`pos` is an internal detail used while producing the query, but is not needed by the caller. This suggests we need an internal object `QueryParserState`:

```
/**
 * @typedef {{
 *     pos: integer,
 *     output: ParsedQuery,
 * }}
 */
var QueryParserState;
```

There may be other fields here that should be internal state.

Reply:
Just like `length`, I moved it out and renamed it `queryPos`.